src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs

```diff
@@ -0,0 +1,99 @@
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+internal partial class Binder
+{
+    /// <summary>
+    /// This visitor walks over a type expression looking for open types.
+    /// 
+    /// Open types are allowed if an only if:
+    /// <list type="number">
+    /// <item>There is no constructed generic type elsewhere in the visited syntax; and</item>
+    /// <item>The open type is not used as a type argument or array/pointer/nullable element type.</item>
+    /// </list>
+    /// 
+    /// Open types can be used both in <c>typeof(...)</c> and <c>nameof(...)</c> expressions.
+    /// </summary>
+    protected sealed class OpenTypeVisitor : CSharpSyntaxVisitor
+    {
+        private Dictionary<GenericNameSyntax, bool>? _allowedMap;
+        private bool _seenConstructed;
+
+        /// <param name="typeSyntax">The argument to typeof.</param>
+        /// <param name="allowedMap">
+        /// Keys are GenericNameSyntax nodes representing unbound generic types.
+        /// Values are false if the node should result in an error and true otherwise.
+        /// </param>
+        public static void Visit(ExpressionSyntax typeSyntax, out Dictionary<GenericNameSyntax, bool>? allowedMap)
+        {
+            OpenTypeVisitor visitor = new OpenTypeVisitor();
+            visitor.Visit(typeSyntax);
+            allowedMap = visitor._allowedMap;
+        }
+
+        public override void VisitGenericName(GenericNameSyntax node)
+        {
+            SeparatedSyntaxList<TypeSyntax> typeArguments = node.TypeArgumentList.Arguments;
+            if (node.IsUnboundGenericName)
+            {
+                _allowedMap ??= [];
+                _allowedMap[node] = !_seenConstructed;
+            }
+            else
+            {
+                _seenConstructed = true;
+                foreach (TypeSyntax arg in typeArguments)
+                {
+                    Visit(arg);
+                }
+            }
+        }
+
+        public override void VisitQualifiedName(QualifiedNameSyntax node)
+        {
+            bool seenConstructedBeforeRight = _seenConstructed;
+
+            // Visit Right first because it's smaller (to make backtracking cheaper).
+            Visit(node.Right);
+
+            bool seenConstructedBeforeLeft = _seenConstructed;
+
+            Visit(node.Left);
+
+            // If the first time we saw a constructed type was in Left, then we need to re-visit Right
+            if (!seenConstructedBeforeRight && !seenConstructedBeforeLeft && _seenConstructed)
+            {
+                Visit(node.Right);
+            }
+        }
+
+        public override void VisitAliasQualifiedName(AliasQualifiedNameSyntax node)
+        {
+            Visit(node.Name);
+        }
+
+        public override void VisitArrayType(ArrayTypeSyntax node)
+        {
+            _seenConstructed = true;
+            Visit(node.ElementType);
+        }
+
+        public override void VisitPointerType(PointerTypeSyntax node)
```

Comment:
>VisitPointerType [](http://example.com/codeflow?start=29&length=16)

Should ```VisitFunctionPointerType``` be overridden as well?

Reply:
Not sure.  @333fred do function pointers need special handling here?

Reply:
note: i've added some typeof and nameof tests taht exercise func-ptrs.  The results seem expected to me.  But i wouldn't mind SMEs here validating.

Reply:
I've played with the change a bit and looked through the logic, and I can't see a scenario where it should matter; since we use `DefaultVisit` for the function pointer, no nested generic types will be marked as allowed. And for any containers, it seems that the following should always hold:

* The container is a generic type, and we won't parse that as an unbound generic type if there are any arguments (ie, `IsUnboundGenericName` won't be true for something like `Dictionary<,delegate*<List<>>`)
* The container is an array or nullable type, and that's always marked as seen constructed.
* The container is a tuple type, and that's never visited by this visitor, so nothing underneath is allowed to be open.

I think that covers all the possibilities, so I don't expect there to be any issues. However, @CyrusNajmabadi, please add a test with 3 generics used like this:
`C<, delegate*<int>, List<>>`

Reply:
@333fred tests added.  test results seemed sensible to me, so i'm moving forward with submitting this.  tnx.